packages/block-editor/src/components/block-list/hover-area.js

```diff
@@ -1,82 +1,41 @@
 /**
  * WordPress dependencies
  */
-import { Component } from '@wordpress/element';
-import { withSelect } from '@wordpress/data';
+import { useState, useEffect } from '@wordpress/element';
 
-class HoverArea extends Component {
-	constructor() {
-		super( ...arguments );
-		this.state = {
-			hoverArea: null,
-		};
-		this.onMouseLeave = this.onMouseLeave.bind( this );
-		this.onMouseMove = this.onMouseMove.bind( this );
-	}
-
-	componentWillUnmount() {
-		if ( this.props.container ) {
-			this.toggleListeners( this.props.container, false );
-		}
-	}
-
-	componentDidMount() {
-		if ( this.props.container ) {
-			this.toggleListeners( this.props.container );
-		}
-	}
-
-	componentDidUpdate( prevProps ) {
-		if ( prevProps.container === this.props.container ) {
-			return;
-		}
-		if ( prevProps.container ) {
-			this.toggleListeners( prevProps.container, false );
-		}
-		if ( this.props.container ) {
-			this.toggleListeners( this.props.container, true );
-		}
-	}
+const useHoveredArea = ( wrapper ) => {
+	const [ hoveredArea, setHoveredArea ] = useState( null );
 
-	toggleListeners( container, shouldListnerToEvents = true ) {
-		const method = shouldListnerToEvents ? 'addEventListener' : 'removeEventListener';
-		container[ method ]( 'mousemove', this.onMouseMove );
-		container[ method ]( 'mouseleave', this.onMouseLeave );
-	}
-
-	onMouseLeave() {
-		if ( this.state.hoverArea ) {
-			this.setState( { hoverArea: null } );
-		}
-	}
+	useEffect( () => {
+		const onMouseLeave = () => {
+			if ( hoveredArea ) {
+				setHoveredArea( null );
+			}
+		};
 
-	onMouseMove( event ) {
-		const { isRTL, container } = this.props;
-		const { width, left, right } = container.getBoundingClientRect();
+		const onMouseMove = ( event ) => {
+			const { width, left, right } = wrapper.current.getBoundingClientRect();
 
-		let hoverArea = null;
-		if ( ( event.clientX - left ) < width / 3 ) {
-			hoverArea = isRTL ? 'right' : 'left';
-		} else if ( ( right - event.clientX ) < width / 3 ) {
-			hoverArea = isRTL ? 'left' : 'right';
-		}
+			let newHoveredArea = null;
+			if ( ( event.clientX - left ) < width / 3 ) {
+				newHoveredArea = 'left';
+			} else if ( ( right - event.clientX ) < width / 3 ) {
+				newHoveredArea = 'right';
+			}
 
-		if ( hoverArea !== this.state.hoverArea ) {
```

Comment:
We're losing this condition in the refactor, which means we're setting state on _every_ `mousemove` event?

Reply:
I thought so as well, but we're not in my tests. I think this is automatically taken care of by react in its `useState` hook. 

Also, adding this test is not a good idea, at least not without moving from `useState` into `useReducer`. When using `useState` it's not safe to use the "previous" state when setting a new one as the previous one might not be the exact previous state since concurrent `setState` calls can happen.

Reply:
> I thought so as well, but we're not in my tests. I think this is automatically taken care of by react in its `useState` hook.

Huh! That's _very_ nice.

For my own curiosity, I found the documented reference to this behavior:

>## Bailing out of a state update
>
>If you update a State Hook to the same value as the current state, React will bail out without rendering the children or firing effects. (React uses the Object.is comparison algorithm.)

https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-state-update